source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,582 @@
+            xConnectInfo.userNameLength = ( uint16_t ) xMQTTUserNameLength;
+            xConnectInfo.keepAliveSeconds = azureiothubKEEP_ALIVE_TIMEOUT_SECONDS;
+            xConnectInfo.passwordLength = ( uint16_t ) ulPasswordLength;
+
+            /* Send MQTT CONNECT packet to broker. Last Will and Testament is not used. */
+            if( ( xMQTTResult = AzureIoTMQTT_Connect( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                      &xConnectInfo,
+                                                      NULL,
+                                                      ulTimeoutMilliseconds,
+                                                      &xSessionPresent ) ) != AzureIoTMQTTSuccess )
+            {
+                AZLogError( ( "Failed to establish MQTT connection: Server=%.*s, MQTTStatus=%d.",
+                              pxAzureIoTHubClient->_internal.ulHostnameLength, pxAzureIoTHubClient->_internal.pucHostname,
+                              xMQTTResult ) );
+                xResult = eAzureIoTHubClientFailed;
+            }
+            else
+            {
+                /* Successfully established a MQTT connection with the broker. */
+                AZLogInfo( ( "An MQTT connection is established with %.*s.", pxAzureIoTHubClient->_internal.ulHostnameLength,
+                             ( const char * ) pxAzureIoTHubClient->_internal.pucHostname ) );
+                xResult = eAzureIoTHubClientSuccess;
+            }
+        }
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_Disconnect( AzureIoTHubClient_t * pxAzureIoTHubClient )
+{
+    AzureIoTMQTTStatus_t xMQTTResult;
+    AzureIoTHubClientResult_t xResult;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to disconnect: Invalid argument." ) );
+        return eAzureIoTHubClientInvalidArgument;
+    }
+
+    if( ( xMQTTResult = AzureIoTMQTT_Disconnect( &( pxAzureIoTHubClient->_internal.xMQTTContext ) ) ) != AzureIoTMQTTSuccess )
+    {
+        AZLogError( ( "Failed to disconnect: %d.", xMQTTResult ) );
+        xResult = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        AZLogInfo( ( "Disconnecting the MQTT connection with %.*s.", pxAzureIoTHubClient->_internal.ulHostnameLength,
+                     ( const char * ) pxAzureIoTHubClient->_internal.pucHostname ) );
+        xResult = eAzureIoTHubClientSuccess;
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                           const uint8_t * pucTelemetryData,
+                                                           uint32_t ulTelemetryDataLength,
+                                                           AzureIoTMessageProperties_t * pxProperties )
```

Comment:
@jspaith Should we add a timeout into this Api, and wait for ack till timeout ? This way we are consistent with out subscribe/unsubscribe api's

Reply:
resolving https://github.com/Azure/azure-iot-middleware-freertos/pull/60#discussion_r607389793 to have the discussion here

Reply:
seems the addition would be here and (in the future) all send API's. (Method response, twin get, etc).

Reply:
other api uses QoS0, we do not need in that.

Reply:
got it so we would add probably a qos option and a timeout in case using qos1?

Reply:
HOw about this.... We have macro in impl for QoS of Telemetry and then in send wait for timeout if QoS1 otherwise return immediately

Reply:
Tracking
https://github.com/Azure/azure-iot-middleware-freertos/issues/80